File: contest/snippets/cpp/graph/prim.cpp
graph prim( graph const& _graph, vertex const _origin )
{
    std::vector< edge > edges;
    std::vector< bool > visited( _graph.size(), false );
    auto cmp_func = []( edge const& _ln, edge const& _rn ) { return std::get<2>(_ln) > std::get<2>(_rn); };
    std::priority_queue< edge, std::vector< edge >, std::function< bool(edge,edge) > > queue( cmp_func );

    visited[_origin] = true;
    for( auto const& kv : _graph[_origin] )
    {
        vertex const v = kv.first;
        cost const c   = kv.second;
        queue.push( edge( _origin, v, c ) );
    }

    while( !queue.empty() )
    {
        edge const e = queue.top();
        queue.pop();

        vertex const from = std::get<1>( e );
        if( visited[from] )
        {
            continue;
        }
        visited[from] = true;

        edges.emplace_back( e );
        for( auto const& n : _graph[from] )
        {
            vertex const to = std::get<0>( n );
            cost const c    = std::get<1>( n );
            queue.push( edge( from, to, c ) );
        }
    }

    graph res = build_graph( _graph.size(), edges );
    return res;
}

File: contest/snippets/python/graph/prim.py
def prim(g, src):
    import heapq

    visited = [False] * len(g)
    spanning_tree_edges = []
    edge_heap = []
    heapq.heappush(edge_heap, (0, (src, src)))
    while edge_heap:
        c, (s, d) = heapq.heappop(edge_heap)
        if visited[d]:
            continue
        visited[d] = True
        spanning_tree_edges.append((s, d, c))

        for _, nd, ndc in g.edges(d):
            heapq.heappush(edge_heap, (ndc, (d, nd)))

    spanning_tree_edges = spanning_tree_edges[1:]
    return Graph(len(g), spanning_tree_edges)
